### PR TITLE
Export import

### LINKED ISSUE #22: Batch import from folder

### LINKED ISSUE #55: Import/Export

### LINKED ISSUE #64: StageCommand should only be available in Storage context

### PR DESCRIPTION
**Pull request by LivingOnEdge**
- Closes #64 Allowed MemeStageCommand to be used in Meme Context
- Closes #55 Implemented working Import Export feature
- Closes #22 Batch import from directory using LoadCommand
- Split up Storage Context into Export and Import Context, and updated UI accordingly.

### CODE REVIEW

File: src/main/java/seedu/weme/logic/parser/WemeParser.java [hunk 1]
```diff
@@ -55,6 +56,9 @@ public Command parseCommand(String userInput) throws ParseException {
         case HelpCommand.COMMAND_WORD:
             return new HelpCommand();
 
+        case MemeStageCommand.COMMAND_WORD:
```
Comment (jonchan51) [inline src/main/java/seedu/weme/logic/parser/WemeParser.java]: This should be in MemeParser right

File: src/main/java/seedu/weme/model/DirectoryPath.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+
+/**
+ * Wrapper class for Path for Import and Export.
+ */
+public class DirectoryPath {
+
+    public static final String MESSAGE_CONSTRAINTS = "File not found or invalid file path given.";
+
+    public final Path filePath;
```
Comment (jonchan51) [inline src/main/java/seedu/weme/model/DirectoryPath.java]: directoryPath would be easier to understand in this case

File: src/main/java/seedu/weme/logic/parser/MemeUnstageCommandParser.java [hunk 3]
```diff
@@ -0,0 +1,28 @@
+package seedu.weme.logic.parser;
+
+import static seedu.weme.commons.core.Messages.MESSAGE_INVALID_COMMAND_FORMAT;
+
+import seedu.weme.commons.core.index.Index;
+import seedu.weme.logic.commands.UnstageCommand;
+import seedu.weme.logic.parser.exceptions.ParseException;
+
+/**
+ * Parses input arguments and creates a new MemeUnstageCommandParser object.
```
Comment (jonchan51) [inline src/main/java/seedu/weme/logic/parser/MemeUnstageCommandParser.java]: MemeUnstageCommand instead of UnstageCommand?

File: src/main/java/seedu/weme/logic/parser/MemeStageCommandParser.java [hunk 4]
```diff
@@ -8,12 +8,12 @@
...
 
 /**
- * MemeStageCommandParser.
+ * Parses input arguments and creates a new MemeStageCommandParser object.
  */
 public class MemeStageCommandParser implements Parser<MemeStageCommand> {
     /**
-     * Parses the given {@code String} of arguments in the context of the MemeDeleteCommand
-     * and returns a MemeDeleteCommand object for execution.
+     * Parses the given {@code String} of arguments in the context of the MemeParseCommand
```
Comment (jonchan51) [inline src/main/java/seedu/weme/logic/parser/MemeStageCommandParser.java]: The comments here don't seem right.

File: src/main/java/seedu/weme/logic/commands/UnstageCommand.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+import seedu.weme.commons.core.Messages;
+import seedu.weme.commons.core.index.Index;
+import seedu.weme.logic.commands.exceptions.CommandException;
+import seedu.weme.model.Model;
+import seedu.weme.model.meme.Meme;
+
+/**
+ * Unstage Command.
+ */
+public class UnstageCommand extends Command {
```
Comment (jonchan51) [inline src/main/java/seedu/weme/logic/commands/UnstageCommand.java]: MemeUnstageCommand? or will this be usable for Templates as well.

File: src/main/java/seedu/weme/logic/commands/MemeStageCommand.java [hunk 6]
```diff
@@ -2,9 +2,13 @@
 
 import static java.util.Objects.requireNonNull;
 
+import java.util.List;
+
+import seedu.weme.commons.core.Messages;
```
Comment (jonchan51) [inline src/main/java/seedu/weme/logic/commands/MemeStageCommand.java]: import the static string instead of the whole class

File: src/main/java/seedu/weme/logic/commands/UnstageCommand.java (hunk 5 above)
Comment (LivingOnEdge) [reply src/main/java/seedu/weme/logic/commands/UnstageCommand.java]: This is only for export tab. The index in unstage 2 refers to the index in the exported list.

File: src/main/java/seedu/weme/logic/parser/MemeExportCommandParser.java [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    /**
+     * Parses the given {@code String} of arguments in the context of the MemeExportCommand
+     * and returns an MemeExportCommand object for execution.
+     * @throws ParseException if the user input does not conform the expected format
+     */
+    public ExportCommand parse(String args) throws ParseException {
+        ArgumentMultimap argMultimap =
+                ArgumentTokenizer.tokenize(args, PREFIX_FILEPATH);
+
+        if (!arePrefixesPresent(argMultimap, PREFIX_FILEPATH)
```
Comment (gary-lgy) [inline src/main/java/seedu/weme/logic/parser/MemeExportCommandParser.java]: Only one prefix is needed here. This can be checked directly without using `arePrefixesPresent`. Even if `arePrefixesPresent` is to be used, it should be pulled out into `ParserUtil` and imported instead of copy/pasted here.

File: src/main/java/seedu/weme/logic/parser/MemeLoadCommandParser.java [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+    }
+
+    /**
+     * Returns true if none of the prefixes contains empty {@code Optional} values in the given
+     * {@code ArgumentMultimap}.
+     */
+    private static boolean arePrefixesPresent(ArgumentMultimap argumentMultimap, Prefix... prefixes) {
+        return Stream.of(prefixes).allMatch(prefix -> argumentMultimap.getValue(prefix).isPresent());
+    }
+}
```
Comment (gary-lgy) [inline src/main/java/seedu/weme/logic/parser/MemeLoadCommandParser.java]: See comments for `MemeExportCommandParser`

File: src/main/java/seedu/weme/model/ModelManager.java [hunk 9]
```diff
@@ -34,6 +34,8 @@
     private final VersionedMemeBook versionedMemeBook;
     private final UserPrefs userPrefs;
     private final FilteredList<Meme> filteredMemes;
+    private final FilteredList<Meme> filteredStagedMemeList;
+    private final FilteredList<Meme> importMemeList;
```
Comment (gary-lgy) [inline src/main/java/seedu/weme/model/ModelManager.java]:
```suggestion
    private final FilteredList<Meme> filteredImportMemeList;
```

File: src/main/java/seedu/weme/ui/MainWindow.java [hunk 10]
```diff
@@ -158,6 +160,15 @@ private void fillAppContent() {
         templatesPanel.getChildren().add(templateGridPanel.getRoot());
 
         // TODO: Fill in other panels here
+
+        MemeGridPanel exportGridPanel = new MemeGridPanel(
+                logic.getFilteredStagedMemeList(), logic.getObservableLikeData());
+        MemeGridPanel importGridPanel = new MemeGridPanel(
+                logic.getImportList(), logic.getObservableLikeData());
```
Comment (gary-lgy) [inline src/main/java/seedu/weme/ui/MainWindow.java]: You're getting like data

File: src/main/java/seedu/weme/logic/commands/UnstageCommand.java (hunk 5 above)
Comment (jonchan51) [reply src/main/java/seedu/weme/logic/commands/UnstageCommand.java]: hmm yeah i dont mind, just leave it be then.

File: src/main/java/seedu/weme/model/MemeBook.java [hunk 11]
```diff
@@ -81,6 +89,56 @@ public boolean hasMeme(Meme meme) {
...
+    }
+
+    /**
+     * Loads meme from given directory to staging area.
+     *
+     * @param directoryPath
+     */
+    public void loadMeme(DirectoryPath directoryPath) throws IOException {
+        // File util importStagingArea
+        FileUtil.load(importList, directoryPath);
```
[the remark this answers is not in the document]
Comment (moziliar) [reply src/main/java/seedu/weme/model/MemeBook.java]: Makes sense.

File: src/main/java/seedu/weme/logic/commands/ExportCommand.java [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+public class ExportCommand extends Command {
+
+    public static final String COMMAND_WORD = "export";
+
+    public static final String MESSAGE_USAGE = COMMAND_WORD
+            + ": Exports memes in the export context to a given directory. "
+            + "Parameters: "
+            + PREFIX_FILEPATH + "PATH \n"
+            + "Example: " + COMMAND_WORD + " "
+            + PREFIX_FILEPATH + "C:/Users/username/Downloads/funny_meme.jpg ";
```
[the remark this answers is not in the document]
Comment (LivingOnEdge) [reply src/main/java/seedu/weme/logic/commands/ExportCommand.java]: Same. If we include that in then the example will be too wordy. Should include in UG instead. Anw if FileUtil is responsible for export so it needs to iterate through a list of memes. Any suggestions? Or should I just generalise the compile time type to List (abit hack).

File: src/main/java/seedu/weme/commons/util/FileUtil.java [hunk 13]
```diff
@@ -121,6 +132,53 @@ public static void copy(InputStream from, Path to) throws IOException {
...
+     * Loads a meme from a given directory into the application.
+     *
+     * @param importList Internal List for memes to be loaded in.
+     * @param directoryPath Path containing memes to load.
+     */
+    public static void load(UniqueMemeList importList, DirectoryPath directoryPath) {
+        final File folder = new File(directoryPath.toString());
+        for (final File fileEntry : folder.listFiles()) {
+            if (fileEntry.isDirectory()) {
+                load(importList, new DirectoryPath(fileEntry.getPath())); // recursive call
```
[the remark this answers is not in the document]
Comment (LivingOnEdge) [reply src/main/java/seedu/weme/commons/util/FileUtil.java]: Cannot due to the need to check for valid image extension. Will consider to change if necessary.

File: src/main/java/seedu/weme/commons/util/FileUtil.java [hunk 14]
```diff
@@ -121,6 +132,53 @@ public static void copy(InputStream from, Path to) throws IOException {
...
+            } else {
+                if (isFileExists(fileEntry.toPath()) && isValidImageExtension(getExtension(fileEntry.toPath()).get())) {
+                    Meme meme = MemeUtil.generateImportMeme(fileEntry.getPath());
+                    importList.add(meme);
```
[the remark this answers is not in the document]
Comment (LivingOnEdge) [reply src/main/java/seedu/weme/commons/util/FileUtil.java]: This is now in the add() method to the UniqueMemeList.

File: src/main/java/seedu/weme/commons/util/FileUtil.java [hunk 15]
```diff
@@ -22,16 +22,92 @@
...
+    }
+
+    /**
+     * Returns a list of valid Image files found in the given directory path.
+     *
+     * @param directoryPath Path containing memes to load.
+     * @return List of loadable paths.
+     */
+    public static List<Path> loadImagePath(DirectoryPath directoryPath) throws IOException {
+        List<Path> pathList = new ArrayList<>();
```
Comment (moziliar) [inline src/main/java/seedu/weme/commons/util/FileUtil.java]: pathList

File: src/main/java/seedu/weme/commons/util/StorageUtil.java [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+    }
+
+    /**
+     * Exports a list of memes to a given directory.
+     *
+     * @param pathList list of memes to be emported.
+     * @param exportPath directory path for the memes to be exported to.
+     * @throws IOException error encountered while exporting.
+     */
+    public static void export(List<Path> pathList, Path exportPath) throws IOException {
```
Comment (moziliar) [inline src/main/java/seedu/weme/commons/util/StorageUtil.java]: This is fine as it exports files in general.

File: src/main/java/seedu/weme/model/MemeBook.java [hunk 17]
```diff
@@ -129,19 +128,17 @@ public void clearExportList() {
     /**
      * Loads meme from given directory to staging area.
      *
-     * @param directoryPath
+     * @param pathList
      */
-    public void loadMemes(DirectoryPath directoryPath) throws IOException {
```
Comment (moziliar) [inline src/main/java/seedu/weme/model/MemeBook.java]: Why this change?

File: src/main/java/seedu/weme/model/ModelManager.java [hunk 18]
```diff
@@ -68,21 +69,24 @@ public ModelManager() {
     //=========== Export/Import ==============================================================================
 
     @Override
-    public void exportMeme(DirectoryPath exportPath) throws IOException {
```
[the remark this answers is not in the document]
Comment (LivingOnEdge) [reply src/main/java/seedu/weme/model/ModelManager.java]: One command is modifying the internal model and the other is retrieving list from the model for external export. Abstract concept doesn't work well here.

[the remark this answers is not in the document]
Comment (LivingOnEdge) [reply src/main/java/seedu/weme/model/ModelManager.java]: This method is handled by ExportCommand.

File: src/main/java/seedu/weme/model/MemeBook.java (hunk 17 above)
Comment (LivingOnEdge) [reply src/main/java/seedu/weme/model/MemeBook.java]: This method is handled by LoadCommand.

File: src/main/java/seedu/weme/model/ModelManager.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (moziliar) [reply src/main/java/seedu/weme/model/ModelManager.java]: Ok got it.

File: src/main/java/seedu/weme/commons/util/FileUtil.java [hunk 19]
```diff
@@ -1,28 +1,123 @@
...
+     * @param imageList
+     * @param directoryPath
+     * @throws IOException
+     */
+    private static void recursiveLoad(List<Path> imageList, DirectoryPath directoryPath) throws IOException {
+        final File folder = toFile(directoryPath);
+        for (final File fileEntry : folder.listFiles()) {
+            if (fileEntry.isDirectory()) {
+                recursiveLoad(imageList, new DirectoryPath(fileEntry.getPath())); // recursive call
+            } else {
```
Comment (jonchan51) [inline src/main/java/seedu/weme/commons/util/FileUtil.java]: just combine this with the if statement

File: src/main/java/seedu/weme/commons/util/FileUtil.java [hunk 20]
```diff
@@ -37,12 +132,26 @@ public static boolean isValidPath(String path) {
...
     }
 
+    /**
+     * Returns true if {@code path} can be converted into a {@code Path} via {@link Paths#get(String)}
+     * and {@link Files#exists(Path)}, otherwise returns false.
+     * @param path A string representing the file path. Cannot be null.
+     */
+    public static boolean isValidDirectoryPath(String path) {
+        try {
+            Paths.get(path);
```
Comment (jonchan51) [inline src/main/java/seedu/weme/commons/util/FileUtil.java]:
```suggestion
            return Files.isDirectory(Paths.get(path));
```

File: src/main/java/seedu/weme/commons/util/FileUtil.java [hunk 21]
```diff
@@ -1,28 +1,123 @@
...
+     *
+     * @param imageList
+     * @param directoryPath
+     * @throws IOException
+     */
+    private static void recursiveLoad(List<Path> imageList, DirectoryPath directoryPath) throws IOException {
+        final File folder = toFile(directoryPath);
+        for (final File fileEntry : folder.listFiles()) {
+            if (fileEntry.isDirectory()) {
+                recursiveLoad(imageList, new DirectoryPath(fileEntry.getPath())); // recursive call
```
Comment (jonchan51) [inline src/main/java/seedu/weme/commons/util/FileUtil.java]: could we just overload the constructor for directorypath and directly pass `fileEntry.toPath()` into directorypath?

File: src/main/java/seedu/weme/commons/util/FileUtil.java (hunk 20 above)
Comment (jonchan51) [reply src/main/java/seedu/weme/commons/util/FileUtil.java]: in fact if you have a path object, you can just directly use Files.isDirectory(path)

File: src/main/java/seedu/weme/logic/commands/ExportCommand.java [hunk 22]
```diff
@@ -0,0 +1,62 @@
...
```
Comment (jonchan51) [inline src/main/java/seedu/weme/logic/commands/ExportCommand.java]: isnt this supposed to be a directory?

File: src/main/java/seedu/weme/model/ModelManager.java [hunk 23]
```diff
@@ -51,13 +54,39 @@ public ModelManager(ReadOnlyWeme weme, ReadOnlyUserPrefs userPrefs) {
...
 
+    //=========== Export/Import ==============================================================================
+
+    @Override
+    public List<Path> getExportPathList() {
+        return versionedWeme.getExportPathList();
+    }
+
+    @Override
+    public void importMeme() throws IOException {
```
Comment (jonchan51) [inline src/main/java/seedu/weme/model/ModelManager.java]:
```suggestion
    public void importMemes() throws IOException {
```
Fix the rest as well

File: src/main/java/seedu/weme/logic/parser/ExportParser.java [hunk 24]
```diff
@@ -0,0 +1,46 @@
...
+import java.util.regex.Matcher;
+
+import seedu.weme.logic.commands.Command;
+import seedu.weme.logic.commands.ExportCommand;
+import seedu.weme.logic.commands.HelpCommand;
+import seedu.weme.logic.commands.UnstageCommand;
+import seedu.weme.logic.parser.exceptions.ParseException;
+
+/**
+ * Parses user input in the import context.
```
Comment (jonchan51) [inline src/main/java/seedu/weme/logic/parser/ExportParser.java]:
```suggestion
 * Parses user input in the export context.
```

File: src/main/java/seedu/weme/logic/parser/MemeUnstageCommandParser.java [hunk 25]
```diff
@@ -0,0 +1,28 @@
...
+import static seedu.weme.commons.core.Messages.MESSAGE_INVALID_COMMAND_FORMAT;
+
+import seedu.weme.commons.core.index.Index;
+import seedu.weme.logic.commands.UnstageCommand;
+import seedu.weme.logic.parser.exceptions.ParseException;
+
+/**
+ * Parses input arguments and creates a new UnstageCommand object.
+ */
+public class MemeUnstageCommandParser implements Parser<UnstageCommand> {
```
Comment (jonchan51) [inline src/main/java/seedu/weme/logic/parser/MemeUnstageCommandParser.java]: Is it `UnstageCommand` or `MemeUnstageCommand`? Fix comments as well

File: src/main/java/seedu/weme/model/Model.java [hunk 26]
```diff
@@ -98,6 +112,18 @@
      */
     void addMeme(Meme meme);
 
+    List<Path> getExportPathList();
```
Comment (jonchan51) [inline src/main/java/seedu/weme/model/Model.java]: Comments

File: src/main/java/seedu/weme/model/Weme.java [hunk 27]
```diff
@@ -93,7 +102,63 @@ public boolean hasMeme(Meme meme) {
...
+
+    public List<Path> getExportPathList() {
+        return exportList.asPathList();
+    }
+
+    /**
+     * Returns true if a template with the same identity as {@code template} exists in the meme book.
+=======
      * Returns true if a template with the same identity as {@code template} exists in weme.
+>>>>>>> master:src/main/java/seedu/weme/model/Weme.java
```
Comment (jonchan51) [inline src/main/java/seedu/weme/model/Weme.java]: this doesnt look like it should be here

File: src/main/java/seedu/weme/ui/ImportMemeCard.java [hunk 28]
```diff
@@ -0,0 +1,70 @@
...
+    @FXML
+    private ImageView display;
+    @FXML
+    private Label id;
+    @FXML
+    private Label description;
+    @FXML
+    private FlowPane tags;
+
+    public ImportMemeCard(Meme meme, int displayedIndex) {
```
Comment (jonchan51) [inline src/main/java/seedu/weme/ui/ImportMemeCard.java]: @moziliar Should we have a BaseMemeCard and a MemeCardWithLikes? This is just memecard without likes

File: src/main/java/seedu/weme/commons/util/FileUtil.java [hunk 29]
```diff
@@ -1,28 +1,123 @@
...
+     * @throws IOException error encountered while copying.
+     */
+    public static void copyFiles(List<Path> pathList, Path folderPath) throws IOException {
+        try {
+            int fileLabel = INITIAL_FILE_LABEL;
+            for (Path path : pathList) {
+                String newFilePath;
+                do {
+                    newFilePath = buildFilePath(folderPath, fileLabel, path);
+                    fileLabel = incrementFileLabel(fileLabel);
```
[the remark this answers is not in the document]
Comment (LivingOnEdge) [reply src/main/java/seedu/weme/commons/util/FileUtil.java]: Better readability for this. I think it is good to have.

File: src/main/java/seedu/weme/commons/util/FileUtil.java (hunk 21 above)
Comment (LivingOnEdge) [reply src/main/java/seedu/weme/commons/util/FileUtil.java]: The code won't necessarily be simpler, another method to validate given path input (currently already have isValidDirectoryPath(String) has to be defined.
